apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py

```diff
@@ -102,70 +165,60 @@ def prepare(element_html, data):
-        element, "grading-method", GRADING_METHOD_DEFAULT
+    grading_method = pl.get_enum_attrib(
+        element, "grading-method", GradingMethodType, GRADING_METHOD_DEFAULT
+    )
+    feedback_type = pl.get_enum_attrib(
+        element, "feedback", FeedbackType, FEEDBACK_DEFAULT
     )
-    feedback_type = pl.get_string_attrib(element, "feedback", FEEDBACK_DEFAULT)
 
-    if grading_method in ["dag", "ranking"]:
-        partial_credit_type = pl.get_string_attrib(element, "partial-credit", "lcs")
-        if partial_credit_type not in ["none", "lcs"]:
-            raise Exception(
-                'partial credit type "'
-                + partial_credit_type
-                + '" is not available with the "'
-                + grading_method
-                + '" grading-method.'
-            )
-    elif pl.get_string_attrib(element, "partial-credit", None) is not None:
+    if (
+        grading_method is not GradingMethodType.DAG
+        and grading_method is not GradingMethodType.RANKING
+        and pl.has_attrib(element, "partial-credit")
+    ):
         raise Exception(
             "You may only specify partial credit options in the DAG and ranking grading modes."
         )
 
-    accepted_grading_method = ["ordered", "unordered", "ranking", "dag", "external"]
-    if grading_method not in accepted_grading_method:
-        raise Exception(
-            "The grading-method attribute must be one of the following: "
-            + ", ".join(accepted_grading_method)
-        )
-
-    if (grading_method not in ["dag", "ranking"] and feedback_type != "none") or (
-        grading_method in ["dag", "ranking"]
-        and feedback_type not in ["none", "first-wrong"]
+    if (
+        grading_method is not GradingMethodType.DAG
+        and grading_method is not GradingMethodType.RANKING
+        and feedback_type is not FeedbackType.NONE
     ):
         raise Exception(
-            'feedback type "'
-            + feedback_type
-            + '" is not available with the "'
-            + grading_method
-            + '" grading-method.'
+            "feedback type {feedback_type.value} is not available with the {grading_method.value} grading-method."
         )
 
-    format = pl.get_string_attrib(element, "format", "default")
+    format = pl.get_enum_attrib(element, "format", FormatType, FormatType.DEFAULT)
     code_language = pl.get_string_attrib(element, "code-language", None)
-    if format != "code" and code_language is not None:
+    if format is FormatType.DEFAULT and code_language is not None:
         raise Exception('code-language attribute may only be used with format="code"')
 
     correct_answers = []
     incorrect_answers = []
     used_tags = set()
 
-    def prepare_tag(html_tags, index, group_info={"tag": None, "depends": None}):
+    def prepare_tag(
+        html_tags: lxml.html.HtmlElement,
+        index: int,
+        group_info={"tag": None, "depends": None},
+    ):
```

Comment:
```suggestion
    ) -> None:
```
It seems like this function is just adding onto the correct answers and incorrect answers lists. However, the loop used could just be moved internally (so the function would instead return these two lists) and this `prepare_tag()` function could live at the global scope (which would simplify the `prepare()` function).

Reply:
I would rather not do that refactor as part of this, I'm trying to stick just to adding types.

Reply:
That's reasonable. Should still set return type to none though.

Reply:
ya, I did